File: packages/php/src/WebhookVerifier.php
<?php

declare(strict_types=1);

namespace Catalisa\Biometrics;

/**
 * Verification of Catalisa Webhooks Engine deliveries:
 *
 *   message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
 *   signature = RSA-SHA256 (PKCS#1 v1.5), base64, in x-webhook-signature as "v1=<base64>"
 *   key       = the subscription's PUBLIC key, chosen by x-webhook-key-id
 *
 * The signature is asymmetric: there is no shared secret to leak. The engine does
 * not refuse old deliveries — the receiver enforces the time tolerance, or a
 * captured delivery could be replayed forever.
 */
final class WebhookVerifier
{
    public const MISSING_HEADERS = 'MISSING_HEADERS';
    public const TIMESTAMP_INVALID = 'TIMESTAMP_INVALID';
    public const TIMESTAMP_OUT_OF_TOLERANCE = 'TIMESTAMP_OUT_OF_TOLERANCE';
    public const UNKNOWN_KEY_ID = 'UNKNOWN_KEY_ID';
    public const UNSUPPORTED_SIGNATURE_VERSION = 'UNSUPPORTED_SIGNATURE_VERSION';
    public const INVALID_SIGNATURE = 'INVALID_SIGNATURE';

    private const MESSAGES = [
        self::MISSING_HEADERS => 'missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature',
        self::TIMESTAMP_INVALID => 'x-webhook-timestamp is not an ISO 8601 date',
        self::TIMESTAMP_OUT_OF_TOLERANCE => 'delivery is outside the time tolerance (possible replay)',
        self::UNKNOWN_KEY_ID => 'no public key for this x-webhook-key-id',
        self::UNSUPPORTED_SIGNATURE_VERSION => 'unsupported signature version (expected v1=)',
        self::INVALID_SIGNATURE => 'invalid signature',
    ];

    /**
     * @param array<string,string> $publicKeys keyId => PEM. Keep a retired key here
     *                                         while deliveries signed with it may still arrive.
     * @param int                  $tolerance  seconds; 0 disables the check (not recommended)
     */
    public function __construct(
        private readonly array $publicKeys,
        private readonly int $tolerance = 300,
        /** Reference clock, for tests. */
        private readonly ?\DateTimeImmutable $now = null,
    ) {
    }

    /**
     * Verifies a delivery and returns the parsed event.
     *
     * Pass the RAW body — the bytes as they arrived. Decoding and re-encoding the
     * JSON changes them, and the signature will not match.
     *
     * @param array<string,string|array<string>> $headers
     * @return array<string,mixed>
     *
     * @throws WebhookVerificationException answer 400 and do not process the body
     */
    public function verify(array $headers, string $rawBody): array
    {
        $id = $this->header($headers, 'x-webhook-id');
        $timestamp = $this->header($headers, 'x-webhook-timestamp');
        $keyId = $this->header($headers, 'x-webhook-key-id');
        $signature = $this->header($headers, 'x-webhook-signature');

        if ($id === null || $timestamp === null || $keyId === null || $signature === null) {
            $this->fail(self::MISSING_HEADERS);
        }

        $sentAt = strtotime($timestamp);
        if ($sentAt === false) {
            $this->fail(self::TIMESTAMP_INVALID);
        }
        if ($this->tolerance > 0) {
            $now = ($this->now ?? new \DateTimeImmutable())->getTimestamp();
            if (abs($now - $sentAt) > $this->tolerance) {
                $this->fail(self::TIMESTAMP_OUT_OF_TOLERANCE);
            }
        }
        if (!str_starts_with($signature, 'v1=')) {
            $this->fail(self::UNSUPPORTED_SIGNATURE_VERSION);
        }
        $pem = $this->publicKeys[$keyId] ?? null;
        if ($pem === null) {
            $this->fail(self::UNKNOWN_KEY_ID);
        }
        $raw = base64_decode(substr($signature, 3), true);
        if ($raw === false) {
            $this->fail(self::INVALID_SIGNATURE);
        }
        $key = openssl_pkey_get_public($pem);
        if ($key === false) {
            $this->fail(self::UNKNOWN_KEY_ID);
        }
        if (openssl_verify("{$id}\n{$timestamp}\n{$rawBody}", $raw, $key, OPENSSL_ALGO_SHA256) !== 1) {
            $this->fail(self::INVALID_SIGNATURE);
        }

        $event = json_decode($rawBody, true);
        if (!is_array($event)) {
            $this->fail(self::INVALID_SIGNATURE, 'the signed body is not valid JSON');
        }

        return $event;
    }

    /** @param array<string,string|array<string>> $headers */
    private function header(array $headers, string $name): ?string
    {
        foreach ($headers as $key => $value) {
            // PHP hands headers over in every shape there is: HTTP_X_WEBHOOK_ID from
            // $_SERVER, x-webhook-id from a framework, arrays from others.
            $normalized = strtolower(str_replace('_', '-', (string) $key));
            $normalized = str_starts_with($normalized, 'http-') ? substr($normalized, 5) : $normalized;
            if ($normalized === $name) {
                $found = is_array($value) ? ($value[0] ?? null) : $value;

                return $found === null || $found === '' ? null : (string) $found;
            }
        }

        return null;
    }

    private function fail(string $failure, ?string $message = null): never
    {
        throw new WebhookVerificationException($failure, $message ?? self::MESSAGES[$failure]);
    }
}

File: packages/java/src/main/java/app/catalisa/biometrics/WebhookVerifier.java
package app.catalisa.biometrics;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.time.Duration;
import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.Base64;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * Verification of Catalisa Webhooks Engine deliveries:
 *
 * <pre>
 * message   = x-webhook-id + "\n" + x-webhook-timestamp + "\n" + raw body
 * signature = RSA-SHA256 (PKCS#1 v1.5), base64, in x-webhook-signature as "v1=&lt;base64&gt;"
 * key       = the subscription's PUBLIC key, chosen by x-webhook-key-id
 * </pre>
 *
 * <p>The signature is asymmetric: there is no shared secret to leak. The engine
 * does not refuse old deliveries — the receiver enforces the time tolerance, or
 * a captured delivery could be replayed forever.
 */
public final class WebhookVerifier {

    /** Which check refused a delivery. */
    public enum Failure {
        MISSING_HEADERS,
        TIMESTAMP_INVALID,
        TIMESTAMP_OUT_OF_TOLERANCE,
        UNKNOWN_KEY_ID,
        UNSUPPORTED_SIGNATURE_VERSION,
        INVALID_SIGNATURE,
    }

    /** A delivery that failed verification. Answer 400 and do not process the body. */
    public static final class VerificationException extends RuntimeException {
        private final Failure failure;

        VerificationException(Failure failure, String message) {
            super(message);
            this.failure = failure;
        }

        public Failure failure() {
            return failure;
        }
    }

    private final Map<String, String> publicKeys;
    private final Duration tolerance;
    private final Supplier<Instant> now;

    /**
     * @param publicKeys keyId to PEM. Keep a retired key here while deliveries
     *                   signed with it may still arrive.
     */
    public WebhookVerifier(Map<String, String> publicKeys) {
        this(publicKeys, Duration.ofMinutes(5), Instant::now);
    }

    /**
     * @param tolerance {@link Duration#ZERO} disables the timestamp check, which
     *                  allows replay — do not.
     */
    public WebhookVerifier(Map<String, String> publicKeys, Duration tolerance, Supplier<Instant> now) {
        this.publicKeys = Map.copyOf(publicKeys);
        this.tolerance = tolerance;
        this.now = now;
    }

    /**
     * Verifies a delivery and returns the parsed event.
     *
     * <p>Pass the RAW body — the bytes as they arrived. Reading the request into
     * a model and re-serializing it changes them, and the signature will not
     * match.
     *
     * @param header reads one header by name, case-insensitively
     */
    public Models.WebhookEvent verify(Function<String, String> header, String rawBody) {
        String id = header.apply("x-webhook-id");
        String timestamp = header.apply("x-webhook-timestamp");
        String keyId = header.apply("x-webhook-key-id");
        String signature = header.apply("x-webhook-signature");

        if (isBlank(id) || isBlank(timestamp) || isBlank(keyId) || isBlank(signature)) {
            throw fail(Failure.MISSING_HEADERS,
                    "missing x-webhook-id, x-webhook-timestamp, x-webhook-key-id or x-webhook-signature");
        }

        Instant sentAt;
        try {
            sentAt = Instant.parse(timestamp);
        } catch (DateTimeParseException e) {
            throw fail(Failure.TIMESTAMP_INVALID, "x-webhook-timestamp is not an ISO 8601 date");
        }
        if (!tolerance.isZero() && Duration.between(sentAt, now.get()).abs().compareTo(tolerance) > 0) {
            throw fail(Failure.TIMESTAMP_OUT_OF_TOLERANCE, "delivery is outside the time tolerance (possible replay)");
        }
        if (!signature.startsWith("v1=")) {
            throw fail(Failure.UNSUPPORTED_SIGNATURE_VERSION, "unsupported signature version (expected v1=)");
        }
        String pem = publicKeys.get(keyId);
        if (pem == null) {
            throw fail(Failure.UNKNOWN_KEY_ID, "no public key for this x-webhook-key-id");
        }

        byte[] raw;
        try {
            raw = Base64.getDecoder().decode(signature.substring(3));
        } catch (IllegalArgumentException e) {
            throw fail(Failure.INVALID_SIGNATURE, "invalid signature");
        }

        boolean valid;
        try {
            PublicKey key = readPublicKey(pem, "RSA");
            Signature verifier = Signature.getInstance("SHA256withRSA");
            verifier.initVerify(key);
            verifier.update((id + "\n" + timestamp + "\n" + rawBody).getBytes(StandardCharsets.UTF_8));
            valid = verifier.verify(raw);
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            throw fail(Failure.UNKNOWN_KEY_ID, "the public key for this keyId is not readable");
        }
        if (!valid) {
            throw fail(Failure.INVALID_SIGNATURE, "invalid signature");
        }

        try {
            return BiometricsClient.MAPPER.readValue(rawBody, Models.WebhookEvent.class);
        } catch (IOException e) {
            throw fail(Failure.INVALID_SIGNATURE, "the signed body is not valid JSON: " + e.getMessage());
        }
    }

    static PublicKey readPublicKey(String pem, String algorithm) throws GeneralSecurityException {
        String body = pem.replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replaceAll("\\s", "");
        byte[] der = Base64.getDecoder().decode(body);

        return KeyFactory.getInstance(algorithm).generatePublic(new X509EncodedKeySpec(der));
    }

    private static boolean isBlank(String value) {
        return value == null || value.isEmpty();
    }

    private static VerificationException fail(Failure failure, String message) {
        return new VerificationException(failure, message);
    }
}
